### DISCUSSION TITLE
Floating-Point Operations and SIMD

### DISCUSSION THREAD

**Discussion by kammt** [kammt/MemeAssembly#63] [Ideas]
Since there is massive demand for a Meme-based programming language capable enough for HPC, I figured why not make MemeAssembly just that.

1. Floating-Point Operations

This would essentially require two things: a) add the possibility to provide floats as immediate arguments and b) make it possible to work with xmm-registers.
For commands, the wisest thing imo would be to just use the commands that already exist. It would essentially require some commands to have a second translation pattern in case floats are used.
Example:
```
xmm1 is brilliant, but I like 420
xmm1 units are ready, with 69 more well on the way
```
should be translated to:
```asm
.section .rodata
.Lf420: .float 420
.Lf69: .float 69

.text
[...]
movss xmm1, [rip + .Lf420]
addss xmm1, [rip + .Lf69]
```
The question here is if and how it should be distinguished between float and double.

3. SIMD-Operations

Now this is where it gets interesting. Implementing this would require
a) 16B Load/Store Instructions - For the sake of simplicity, I would for now only use unaligned Load/Store Instructions (because I mean people that really want their code to be fast use `-O69420` anyway)
b) all arithmetic operations to support a packed version of itself

The second part is a bit problematic. First of all, how should a user differentiate a packed from a non-packed instruction (i.e. what prefix/suffix should be used?)? 
Secondly, SIMD does not just cover floats, but also shorts, ints etc., so how would go about setting what exact instruction they want? (It would be possible though to just restrict it to floats only)
Thirdly, how would you implement comparisons in a way that makes sense? 
Example: `corparate needs you to find the difference between xmm1 and 2` as a packed instruction: Some values might be equal to 2, some might not. So under which conditions should the program then jump to `they're the same picture`? When every value is equal?  If we take this route, how should `who would win?` be implemented, since there is currently no possibility for a "none of the above option" i.e. not all values are equal, above or below the second one.

I'll just leave this here and post ideas as I get them - feedback is always welcome!

**Comment by jmrtsh** [comment]
I think trying to automatically map the comprehensive list of x86-64 SIMD Instructions to the same meme mnemonics that are used in the current integer operations would be quite a challenge and amount to creating an entire meta SIMD programming language, rather than sticking to the original idea of assembly where one instruction does one thing and nothing else ([damn you `mov`](https://harrisonwl.github.io/assets/courses/malware/spring2017/papers/mov-is-turing-complete.pdf)). 

And to answer b):  I would propose two modes of operation instead of a prefix/suffix syntax, something like `I am once again asking for your floating point`, not as an instruction per se, but as something of a compiler pragma, just hidden as normal code, to guarantee the lowest readability possible :). After this pseduo instruction has been called all operations such as `xmm1 is brilliant, but I like 42` would implicitly be floating point loads/stores/arithmetic/comparisons/etc. .A similar approach could be used to distinguish between single and double precision operations, maybe a 'pragma' such as [`Speed and Power`](https://www.youtube.com/watch?v=ToyM657PxYk) could switch between single and double precision instructions? But I am less certain about this kind of mode switching for single and double precision as those two are quite often used in tandem and permanent switching could be annoying (well except for the numerous `speed and power` lines in the code, those would be quite fun).
I'm uncertain about comparisons though, they might prove quite hard, the problem of distinguishing between two kinds of x86-64 SIMD comparisons, the classic flag setting instructions such as the endless `[u]comis[s]` variantions as well as the `pcmp[eq/gt][b/w/d/q]` instructions the latter of which are often more elegant and performant, so one could also consider just disregarding the 'comparison by flag' instructions for floats entirely.

**Reply by kammt** [reply]
"Operation modes" instead of using prefixes is a good idea! I would propose using `Speed and Power` for switching between SIMD and SISD, and `Prepare for trouble and make it double` for switching between float/double. If SIMD is set and floats are disabled, 64bit or 32Bit (long/int) could be used.

Examples:
```
I am once again asking for your floating point
Speed and Power
```
Sets the mode to SIMD floats. You can also use this for an extra dose of code obfuscation and do something like

```
I am once again asking for your floating point
Speed and Power
Prepare for trouble and make it double
Speed and Power
I am once again asking for your floating point
Speed and Power
```
to set the mode to SIMD int

Now to comparisons:
- SISD float/double: I would use `ucomis[s/d]` for this.
- SIMD float/double/int: In my opinion, this would be the easiest implementation:
   - Check equality: `corporate needs you to find the difference...` compares both xmm registers for equality, then jumps to `they're the same picture` if *all* lanes are equal. If not all lines are equal, no jump is performed and the destination register is masked
   - Check greater/lower than: `who would win?` Jumps to the respective label if *all* lanes are greater than / lower than the other lane. If this is not the case, no jump will be performed and the destination register is masked to indicate the result.
The alternative would be to introduce completely new commands for comparisons, which would be more confusing in my opinion (And I cannot think of any other meme that could be used for comparisons)
